crates/ruff_python_formatter/src/expression/expr_tuple.rs

```diff
@@ -1,17 +1,130 @@
+use crate::context::PyFormatContext;
 use crate::expression::parentheses::{
     default_expression_needs_parentheses, NeedsParentheses, Parentheses, Parenthesize,
 };
-use crate::{not_yet_implemented_custom_text, FormatNodeRule, PyFormatter};
-use ruff_formatter::{write, Buffer, FormatResult};
+use crate::trivia::{first_non_trivia_token, TokenKind};
+use crate::{AsFormat, FormatNodeRule, FormattedIterExt, PyFormatter, USE_MAGIC_TRAILING_COMMA};
+use ruff_formatter::formatter::Formatter;
+use ruff_formatter::prelude::{
+    block_indent, group, hard_line_break, if_group_breaks, soft_block_indent, soft_line_break,
+    soft_line_break_or_space, text,
+};
+use ruff_formatter::{format_args, write, Buffer, Format, FormatResult};
+use ruff_python_ast::prelude::{Expr, Ranged};
+use ruff_text_size::TextRange;
+
 use rustpython_parser::ast::ExprTuple;
 
 #[derive(Default)]
 pub struct FormatExprTuple;
 
 impl FormatNodeRule<ExprTuple> for FormatExprTuple {
-    fn fmt_fields(&self, _item: &ExprTuple, f: &mut PyFormatter) -> FormatResult<()> {
-        write!(f, [not_yet_implemented_custom_text("(1, 2)")])
+    fn fmt_fields(&self, item: &ExprTuple, f: &mut PyFormatter) -> FormatResult<()> {
+        let ExprTuple {
+            range,
+            elts,
+            ctx: _,
+        } = item;
+
+        // Handle the edge cases of an empty tuple and a tuple with one element
+        let last = match &elts[..] {
+            [] => {
+                return write!(
+                    f,
+                    [group(&format_args![
+                        // A single element tuple always needs parentheses
+                        &text("("),
+                        block_indent(&dangling_node_comments(item)),
+                        &text(")"),
+                    ])]
+                );
+            }
+            [single] => {
+                return write!(
+                    f,
+                    [group(&format_args![
+                        // A single element tuple always needs parentheses
+                        &text("("),
+                        soft_block_indent(&format_args![single.format(), &text(",")]),
+                        &text(")"),
+                    ])]
+                );
+            }
+            [.., last] => last,
+        };
+
+        let magic_trailing_comma = USE_MAGIC_TRAILING_COMMA
+            && first_non_trivia_token(last.range().end(), f.context().contents())
+                .map(|token| token.kind)
+                == Some(TokenKind::Comma);
+
+        if magic_trailing_comma {
+            // A magic trailing comma forces us to print in expanded mode since we have more than
+            // one element
+            write!(
+                f,
+                [
+                    // An expanded group always needs parentheses
+                    &text("("),
+                    hard_line_break(),
+                    block_indent(&ExprSequence::new(elts)),
+                    hard_line_break(),
```

Comment:
The `hard_line_breaks` are not necessary. The `block_indent` adds the line breaks for you (unlike `indent` that never renders line breaks and relies on you to add the appropriate line breaks)

```suggestion
                    block_indent(&ExprSequence::new(elts)),
```

You can't see the extra `hard_line_breaks` in the output because the `Printer` optimizes them away.